common/src/primitives/amount.rs

```diff
@@ -317,6 +317,35 @@ mod tests {
         assert_eq!(Amount { val: IntType::MAX } + Amount { val: 1 }, None);
     }
 
+    #[test]
+    fn sum_some() {
+        let amounts = vec![Amount { val: 1 }, Amount { val: 2 }, Amount { val: 3 }];
+        assert_eq!(
+            amounts.into_iter().sum::<Option<Amount>>(),
+            Some(Amount { val: 6 })
+        );
+    }
+
+    #[test]
+    fn sum_overflow() {
+        let amounts = vec![
+            Amount { val: 1 },
+            Amount { val: 2 },
+            Amount {
+                val: IntType::MAX - 2,
+            },
+        ];
+        assert_eq!(amounts.into_iter().sum::<Option<Amount>>(), None);
+    }
+
+    #[test]
+    fn sum_empty() {
+        assert_eq!(
+            vec![].into_iter().sum::<Option<Amount>>(),
+            Some(Amount::from_atoms(0))
```

Comment:
I think this answer is controversial. This is why I asked for a test for the empty vector. Shouldn't this be `None`? We're adding nothing... so we get nothing.

Reply:
I don't think this is controversial.
In mathematics, the sum of an empty collection of elements is zero.

Reply:
But there's a difference in math. The sum of an empty set is zero, assuming the elements of the set have a numeric type. That's the difference here. So, if it's `Vec<u32>`, then sure, the sum of an empty `Vec<u32>`, is definitely zero. But then what's the sum of `Vec<Option<u32>>` when it's empty? I don't think that can be decided without a convention (which is not the case with numerical types). What do you think @iljakuklic ?

Reply:
I think `Some(0)` is the right result here. Summing an empty vector should give the neutral element, i.e. one where `0 + x = x + 0 = x` for all `x` (where `0` is the neutral element). Then we have some nice properties like monoid homomorphism: `concat(xs, ys).sum() = xs.sum() + ys.sum()`.

So let's define `+` on `Option<int>`:

```rust
fn add(x, y: Option<IntType>) {
    match (x, y) {
        (Some(x), Some(y)) => Some(x + y),
        (_, _) => None,
    }
}
```

This is consistent with how summing a vector of amounts behaves for the non-empty case. The element that acts as a neutral element for the above operation is `Some(0)`.

Reply:
Sounds fair. Let's go for it. Cheers!